### PR TITLE
adds watch-list implementation without breaking changes

### PR DESCRIPTION
**Pull request by casualjim**
## Motivation

Implement support for [watch-list](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/3157-watch-list/README.md). 

We have 18000 namespaces in our cluster, we store custom resources where some collections have large schemas and the cardinality is also in the 100k's. For example, when a client doesn't use pagination our API server memory usage can balloon up to 2-300 GB under certain conditions

## Solution

The most important goal is to reduce memory pressure on the API server. It would be nice to be able to reduce it on the client side too, in that case, we also care about knowing when the initial sync and or resync is in progress or at the very least completed. 

Since this is for a new k8s version, I think it might be ok to change the watch event contract slightly.  I've come up with a couple of approaches

Option 1: do nothing, still accumulate everything in a vec and emit `Event::Restarted` as the first event, when this streaming model is enabled. The name of the event is a bit weird, and it could maybe be renamed to express intent better although that would introduce breaking changes
Option 2: Modify the applied event to have an option<bool> or an enum value to indicate if we're still in the initial listing state
Option 3: Introduce 2 new events, one that indicates a relist is starting and one that indicates the relist is completed, this can then remove the need for the Restarted event because these 2 new events give enough signal. 
Option 4: stop being helpful and just expose the same events as the plain k8s API

Personally I'm in favor of option 3. Option 3 will allow for keeping client-side memory usage low, and it would allow for flagging a store as in or out of sync.
This PR currently holds option 1

This is a draft, so it's not signed. I'll work through what is required to submit PR's from my work email for the final PR. I just wanted to gather feedback.

closes: #1254

### CODE REVIEW

**Comment by clux** [comment]
Hey, thanks a lot for this, and while this is going further than the original help-wanted issue (as the runtime is a little less clear), I appreciate the taking the initiative in an open-minded way with several options. I'll leave some comments on the PR, but some high level comments first:

Personally, I am leaning towards option 3 also for the `watcher` side of things just as that's one of the few things that can actually help with the client spikiness. An option 3 would necessarily involve buffering at the reflector side, but that would actually help people both using reflectors and people just watching:

- `watcher` users get everything __as it comes in__ 
- `reflector` users can prune before caching (using `WatchStreamExt::modify`) because the raw events are passed through from watcher before it is buffered - so it is still a spike, but it's a more controllable spike

Note also that in general the aim is not to abstract away the k8s api. Particularly, kube-core and kube-client should just do the k8s api (where copy-pasted doc-strings are often a thing for the same parameters).  So i don't think 3 and 4 are necessarily in conflict.

File: kube-runtime/src/watcher.rs
Comment (clux) [inline kube-runtime/src/watcher.rs:335]:
```suggestion
    pub fn streaming_lists(mut self) -> Self {
        self.bookmarks = true; // streaming lists require bookmarks
        self.streaming = WatcherMode::StreamingList;
        self.page_size = None; // page size does not make sense when streaming
```

..it might also be worth setting
`self.list_semantic(ListSemantic::Any)` in this fn. it almost seems like the [streaming list docs](https://kubernetes.io/docs/reference/using-api/api-concepts/#streaming-lists) are recommending it, and it might be a good opt-in point for users (but not sure about this)

**Review by clux — COMMENTED** [review]
quick points

**Comment by casualjim** [comment]
I was thinking that because it's an API change, we could create a watcher2 or streaming_watcher, that has the event contract from option 3. That way we can mark the previous implementation as deprecated, and carry it forward a few versions before removing it. That would give people who depend on this library time to move to the new implementation instead of it being forced on them.

File: kube-runtime/src/watcher.rs
Comment (nightkr) [reply kube-runtime/src/watcher.rs:335]: Also, I'd probably move `page_size` into `WatcherMode`, but I suppose our hands are tied here...

Comment (nightkr) [reply kube-runtime/src/watcher.rs:335]: Third, `streaming` makes it sound like it should be a boolean, how about `watcher_mode` (like the enum)?

**Comment by casualjim** [comment]
I've addressed the comments

File: kube-core/src/params.rs
Comment (clux) [inline kube-core/src/params.rs:340]: This tries to change the existing resource version that is normally [passed out-of-band](https://docs.rs/kube/latest/kube/core/struct.Request.html#method.watch), and does not look like it is serialized.

It is perhaps a bit strange and arguable that we are passing the resource version out of band outside the watchparams, but it does allow repeat calls more easily without having to constantly mutate the params, and i don't think it needs to be changed here.

File: kube-core/src/params.rs
Comment (clux) [inline kube-core/src/params.rs:335]: These flags need serialization tests in request.rs in kube-core to ensure they are serialized into the query params.

File: kube-core/src/params.rs
Comment (clux) [inline kube-core/src/params.rs:498]: Links up with previous comment. Does not actually do anything.

File: kube-runtime/src/watcher.rs
Comment (clux) [inline kube-runtime/src/watcher.rs:243]: This comment seems a bit wrong, it's written as if users have already chosen streaming lists, when that's an option in the enum.

**Review by clux — CHANGES_REQUESTED** [review]
General setup looks good! Some issues with resource versions, missing version gates, and partially bumped k8s versions. It's also currently not working on the 1.27 cluster i tested this on:

```
2023-07-22T21:10:53.985375Z DEBUG kube_runtime::utils::stream_backoff: Error received, backing off deadline=Instant { tv_sec: 37803, tv_nsec: 770027883 } duration=85.857327ms
Error: watch stream failed: ApiError: ListOptions.meta.k8s.io "" is invalid: resourceVersionMatch: Forbidden: resourceVersionMatch is forbidden for watch unless sendInitialEvents is provided: Invalid (ErrorResponse { status: "Failure", message: "ListOptions.meta.k8s.io \"\" is invalid: resourceVersionMatch: Forbidden: resourceVersionMatch is forbidden for watch unless sendInitialEvents is provided", reason: "Invalid", code: 422 })

Caused by:
    0: ApiError: ListOptions.meta.k8s.io "" is invalid: resourceVersionMatch: Forbidden: resourceVersionMatch is forbidden for watch unless sendInitialEvents is provided: Invalid (ErrorResponse { status: "Failure", message: "ListOptions.meta.k8s.io \"\" is invalid: resourceVersionMatch: Forbidden: resourceVersionMatch is forbidden for watch unless sendInitialEvents is provided", reason: "Invalid", code: 422 })
    1: ListOptions.meta.k8s.io "" is invalid: resourceVersionMatch: Forbidden: resourceVersionMatch is forbidden for watch unless sendInitialEvents is provided: Invalid
```

when using
```diff
diff --git a/examples/pod_watcher.rs b/examples/pod_watcher.rs
index e4e79fc4..a485892f 100644
--- a/examples/pod_watcher.rs
+++ b/examples/pod_watcher.rs
@@ -13,7 +13,7 @@ async fn main() -> anyhow::Result<()> {
     let client = Client::try_default().await?;
     let api = Api::<Pod>::default_namespaced(client);
 
-    watcher(api, watcher::Config::default())
+    watcher(api, watcher::Config::default().streaming_lists().any_semantic())
         .applied_objects()
         .default_backoff()
         .try_for_each(|p| async move {
```

File: kube-runtime/src/watcher.rs
Comment (clux) [reply kube-runtime/src/watcher.rs]: My original advice above was indeed misguided. However, one point; you do need to set the `version_match` parameter for `WatchParams` in this mode or you'll literally get a `Error: failed to start watching object: Failed to build request: failed to validate request: WatchParams::version_match must be set to NotOlderThan when using send_initial_events` if you don't.

We don't really have much chance as to what to set the resource version to (it's either "0" or an explicit from previous stored version) so it means that it's a mistake to take [list-semantics](https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list) and try to re-use them for [streaming-list semantics](https://kubernetes.io/docs/reference/using-api/api-concepts/#streaming-lists) because the latter only supports `NotOlderThan` semantics. The `ListSemantic` option was only ever used to produce `ListParams` so we shouldn't use it to create `WatchParams`.

If we are in `streaming_list` mode, then we must set `WatchParams::version_match` to `NotOlderThan`, I don't think there are any other ways to shoehorning this in.

Comment (clux) [reply kube-runtime/src/watcher.rs]: As for the `page_size.is_some() && WatchSemantic::InitialEvents` scenario, we could maybe add a `Config::validate` method that is called by the watcher before starting? That is inline with our other *Params structs.

File: kube-core/src/params.rs
Comment (casualjim) [reply kube-core/src/params.rs:340]: I'll remove this one, it's handled differently anyway.

**Comment by casualjim** [comment]
I'll add a namespace example reflector that shows the watch list behavior. Do I feature gate all the doc comments I made?

**Comment by casualjim** [comment]
I've added an example now and updated the k3d configuration that enables the feature gate

File: kube-core/src/params.rs
Comment (clux) [reply kube-core/src/params.rs]: Seeing the new tests, looking good :+1:

File: kube-core/src/params.rs
Comment (clux) [inline kube-core/src/params.rs:387]: This is a sensible default for now. I think we should maybe add a new constructor that sets correct flags rather than mess with the builders/serialization. Something like:

```rust
impl WatchParams
    /// Constructor for doing Kubernetes 1.27 Streaming List watches
    ///
    /// Enables [`VersionMatch::NotGreaterThan`] semantics and [`WatchParams::send_initial_events`].
    pub fn streaming_lists() -> Self {
        Self {
            send_initial_events: true,
            version_match: Some(VersionMatch::NotGreaterThan), // required
            bookmarks: true, // required
            ..WatchParams::default()
        }
    }
```

File: kube-runtime/src/watcher.rs
Comment (clux) [inline kube-runtime/src/watcher.rs:235]: should caveat that options such as this do not always have an effect anymore:

```suggestion
    /// Configures re-list for performance vs. consistency.
    ///
    /// NB: This option only has an effect for [`WatcherMode::ListWatch`].
    pub list_semantic: ListSemantic,
```

we probably should also add this to `::page_size` plus the builders `::any_semantic()`, `::list_semantic()`, `::page_size()`.

File: kube-runtime/src/watcher.rs
Comment (clux) [inline kube-runtime/src/watcher.rs:249]:
```suggestion

    /// Control how the watcher fetches the initial list of objects.
    ///
    /// ListWatch: The watcher will fetch the initial list of objects using a list call.
    /// StreamingList: The watcher will fetch the initial list of objects using a watch call.
    ///
    /// StreamingList is more efficient than ListWatch, but it requires the server to support
    /// streaming list bookmarks (opt-in feature gate in Kubernetes 1.27).
    ///
    /// See https://kubernetes.io/docs/reference/using-api/api-concepts/#streaming-lists
    /// See https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3157-watch-list#design-details
```

Reworded a bit. Plus removed the line about fallback because if the server does not support it, it just errors.

```
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
     Running `/home/clux/kube/kube/target/debug/examples/namespace_reflector`
Error: watch stream failed: ApiError: ListOptions.meta.k8s.io "" is invalid: sendInitialEvents: Forbidden: sendInitialEvents is forbidden for watch unless the WatchList feature gate is enabled: Invalid (ErrorResponse { status: "Failure", message: "ListOptions.meta.k8s.io \"\" is invalid: sendInitialEvents: Forbidden: sendInitialEvents is forbidden for watch unless the WatchList feature gate is enabled", reason: "Invalid", code: 422 })

Caused by:
    0: ApiError: ListOptions.meta.k8s.io "" is invalid: sendInitialEvents: Forbidden: sendInitialEvents is forbidden for watch unless the WatchList feature gate is enabled: Invalid (ErrorResponse { status: "Failure", message: "ListOptions.meta.k8s.io \"\" is invalid: sendInitialEvents: Forbidden: sendInitialEvents is forbidden for watch unless the WatchList feature gate is enabled", reason: "Invalid", code: 422 })
    1: ListOptions.meta.k8s.io "" is invalid: sendInitialEvents: Forbidden: sendInitialEvents is forbidden for watch unless the WatchList feature gate is enabled: Invalid
```

File: kube-core/src/params.rs
Comment (clux) [inline kube-core/src/params.rs:332]: client-go might be defaulting this to true right away, but they are on a much stricter version skew policy. 

```suggestion
```

Sidenote: This all reads heavily as "we don't want you to ask for consistent reads" so we are changing the default behaviour going forward.

**Review by clux — COMMENTED** [review]
This does actually work for me now! Thanks for the k3d arg.

Left a few more suggestions on serialization/builders/configuration docs. This is veering a bit more into the whole philosophical discussion of "making invalid states unrepresentable" (which is something we want to do for the abstractions, but we not necessarily something we sensible can completely disallow __at the serialization layer__ because we don't necessarily know what comes next - or if Kubernetes' docs are correct/will stay correct).

Have in general suggested moving our own safety checks and preconfigured "good defaults" into validators + and added some suggestions for `watcher::Config` to make the correct choice more clear, and less implicit.

(There may be some follow-up issues that are worth raising here on our end as well; maybe we should not allow disabling bookmarks in watcher, and maybe we shouldn't be hard-enforcing a timeout of 295s in validate.)

File: kube-core/src/params.rs
Comment (casualjim) [reply kube-core/src/params.rs:350]: I removed this option from the watch params. I figured there is no use case for making in configurable on the watch params for now as it's implicit. Similar to how they are treated in the list semantics path

File: kube-runtime/src/watcher.rs
Comment (casualjim) [inline kube-runtime/src/watcher.rs:204]: I renamed this one because clippy is unhappy abou the stuttering of package name and type name `watcher::WatcherMode`

File: kube-runtime/src/watcher.rs
Comment (nightkr) [reply kube-runtime/src/watcher.rs]: I'd want to at least flag this to the user somehow as something to be aware of...

**Comment by Arnavion** [comment]
>There's nothing really blocking from the `k8s-openapi` side to release their new minor with `v1_27` (as we have both tested that it works)

Cool. I'll release it this weekend then.

**Comment by casualjim** [comment]
I would like to get past this hurdle of CI yes. Unfortunately I have to do planning and roadmap work for my project (this library is part of said project) and will need to spend a lot of time in google docs instead of in code this weekend. The failures are about building k8s-openapi so I'm going to wait for a release on that.

I think the tests etc passed when I ran them locally so hopefully this will build here too. If not I can take a look again next weekend perhaps.

File: e2e/Cargo.toml
Comment (Arnavion) [reply e2e/Cargo.toml:21]: FYI You can do `"k8s-openapi/latest"` now.

**Comment by casualjim** [comment]
I don't know what to do with CI. 
I can run the just tasks just fine on my computer but as soon as I try to do it with `act` the deny starts failing and something that's supposed to set up k3d keeps telling me that `kubectl` can't be found.

So it's difficult to get this to work for me atm. This could use some help from somebody who knows the entire toolchain

File: Cargo.toml
Comment (clux) [inline Cargo.toml:3]: i would remove this unless you had a reason for it, i vaguely remember it caused problems in the past.

**Comment by clux** [comment]
> I don't know what to do with CI. I can run the just tasks just fine on my computer but as soon as I try to do it with `act` the deny starts failing and something that's supposed to set up k3d keeps telling me that `kubectl` can't be found.

I think a lot of the CI failures atm is due to resolver (because it's getting the multiple `k8s-openapi` versions enabled message). Don't worry about the deny failure, will fix that in main (related to us not maintaining a lockfile).

**Comment by casualjim** [comment]
I undid the resolver = 2 thing, hopefully that makes it go further. 

When I run locally most things seem to succeed but there is one thing that can not find kubectl and it will just hang on that

File: kube-core/src/params.rs
Comment (clux) [reply kube-core/src/params.rs:368]: resolving this, we are going in the other direction than what i expected in https://github.com/kube-rs/kube/pull/1281 so what you have here is matching that.

File: kube-runtime/src/watcher.rs
Comment (clux) [reply kube-runtime/src/watcher.rs]: In a similar vein to the other resolved thread, the `VersionMatch` is set in kube-core in the streaming lists case as [is required](https://kubernetes.io/docs/reference/using-api/api-concepts/#streaming-lists) and this matches the strategy taken in https://github.com/kube-rs/kube/pull/1281/files

File: kube-runtime/src/watcher.rs
Comment (clux) [reply kube-runtime/src/watcher.rs:251]: This comment suggestion, or something like it should still be applied. Currently this comment is on the `initial_list_strategy` member, and that does not need to be configured to use streaming lists (as that only refers to one variant).

**Comment by clux** [comment]
Had another look over of this and did another test, and this seems fine to me. Resolved a few comments as per strategy in https://github.com/kube-rs/kube/pull/1281/files (which is nice to be able to justify as it does make invalid states unrepresentable).

One doc nit remains (config member doc suggestion), and the other two threads are just notes for the future. I am otherwise happy with this. Easy to opt-in, hard to do the wrong thing, works well, CI now green. Clean addition. Thanks a lot for doing this!

**Comment by casualjim** [comment]
I updated the doc and added that extra constructor method

**Comment by clux** [comment]
Thanks, looks good! But needs the DCO signoff again now.

**Review by clux — APPROVED** [review]

**Pull request merged**
